Cython/Compiler/ExprNodes.py

```diff
@@ -2042,11 +2042,12 @@ def infer_type(self, env):
                 return self.inferred_type
             return py_object_type
         elif (self.entry.type.is_extension_type or self.entry.type.is_builtin_type) and \
-                not self.is_target and self.name == self.entry.type.name:
+                self.entry == self.entry.type.entry:
```

Comment:
In which case is `entry != entry.type.entry` ? (It makes a difference, so there probably are such cases.)

Reply:
`self.entry == self.entry.type.entry` means "the entry is the same one that defines the type"

so in

```
str = ""
str  # the entry for `str` is no longer the "type.entry" - it's just an entry in the local scope
```

Reply:
Right, that makes total sense.